pkg/csi/service/wcp/controller.go

```diff
@@ -1237,149 +1239,228 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 		}
 		volumeType = prometheus.PrometheusBlockVolumeType
 
-		var isVADeleted bool
-		volumeAttachment, err := commonco.ContainerOrchestratorUtility.GetVolumeAttachment(ctx, req.VolumeId, req.NodeId)
-		if err != nil {
-			if apierrors.IsNotFound(err) {
-				log.Infof("VolumeAttachments object not found for volume %q & node %q. "+
-					"Thus, assuming the volume is detached.", req.VolumeId, req.NodeId)
-				isVADeleted = true
-			} else {
-				log.Errorf("failed to retrieve volumeAttachment from API server Err: %v", err)
-				return nil, csifault.CSIInternalFault, err
+		csiDetachEnabled := commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx,
+			common.CSIDetachOnSupervisor)
+
+		if csiDetachEnabled {
+			var isVADeleted bool
+			var vmuuid string
+			volumeAttachment, err := commonco.ContainerOrchestratorUtility.GetVolumeAttachment(ctx, req.VolumeId, req.NodeId)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					log.Infof("VolumeAttachments object not found for volume %q & node %q. "+
+						"Thus, assuming the volume is detached.", req.VolumeId, req.NodeId)
+					isVADeleted = true
+				} else {
+					log.Errorf("failed to retrieve volumeAttachment from API server Err: %v", err)
+					return nil, csifault.CSIInternalFault, err
+				}
 			}
-		}
-		if !isVADeleted {
-			log.Debugf("controllerUnpublishVolumeInternal: VA : %v", spew.Sdump(volumeAttachment))
-			for k, v := range volumeAttachment.Status.AttachmentMetadata {
-				if k == common.AttributeVmUUID {
-					log.Debugf("controllerUnpublishVolumeInternal: vmuuid value: %q", v)
-					vcdcMap, err := getDatacenterFromConfig(c.manager.CnsConfig)
-					if err != nil {
-						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-							"failed to get datacenter from config with error: %+v", err)
+			if !isVADeleted {
+				log.Debugf("controllerUnpublishVolumeInternal: VA : %v", spew.Sdump(volumeAttachment))
+				for k, v := range volumeAttachment.Status.AttachmentMetadata {
+					if k == common.AttributeVmUUID {
```

Comment:
do we break once we found `k == common.AttributeVmUUID` and processed it.

Reply:
I can add a break. AttachmentMetadata is a map and `vmUUID` is the key we are looking for. There will only be 1 key like that in the map.